### ISSUE TITLE
RTL text not working since Eclipse 4.17 (Archi 4.8+)

### ISSUE THREAD

**Issue by Alon-fnx** [archimatetool/archi#796]
Bug report

### Version of Archi
4.9.1

### Archi Plug-ins
Collaboration

### Operating System
Win10

### Expected Behaviour
Text in shape should support RTL text

### Actual Behaviour
Words written in Hebrew in shapes are written in LTR order
For example, If I type
שלום עולם
It shows:
עולם שלום

### Steps to Reproduce the Behaviour
1. Edit a diagram shape
2. Enter a sentence in Hebrew
3. The shape shows the words in opposite order

**Comment by Alon-fnx** [comment]
![1](https://user-images.githubusercontent.com/97157181/148190304-f735704d-c60b-45eb-b652-e5bbb2e737de.jpg)
![2](https://user-images.githubusercontent.com/97157181/148190315-c1c97921-dfe5-4a14-8789-b2b3929ee1e4.jpg)

**Comment by Alon-fnx** [comment]
Sשצפךקד

**Comment by Phillipus** [comment]
I don't know what the level of support for RTL text is in Eclipse based apps. I Googled the issue and found that one can add the following two lines to the `Archi.ini` file:

```
-dir
rtl
``

(Apparently RTL text is not supported in the StyledText component.)

As this is an open source project please feel free to investigate the issue further (Google search is a good start)

**Comment by Alon-fnx** [comment]
Thanks for the fast response.

The strange thing is that in previous versions, this problem did not exist...
I'll try and notify.

**Comment by Phillipus** [comment]
> The strange thing is that in previous versions, this problem did not exist...

Is RTL working in, for example, a single line text control like "Name"?

Things that changed between Archi 4.8.1 and 4.9.x:

Eclipse 4.18 -> Eclipse 4.21
We forked the Eclipse GEF and Draw2d frameworks

You could try Archi 4.9.2 beta (uses Eclipse 4.22) https://www.archimatetool.com/beta/

**Comment by Alon-fnx** [comment]
The problem is happening even in a single line of text - as shown in the example.
I'm talking about versions 4.4 / 4.7 and before

I'll try the beta

**Comment by Phillipus** [comment]
I'm not asking about the text in a diagram figure. I mean single line text controls as in the Properties tab or the controls in the Preferences dialog.

You could also test against Archi 4.8.1 - https://www.archimatetool.com/archive/

**Comment by Alon-fnx** [comment]
In any other object (Properties, diagram name etc'.) it's ok.
It happens only in diagram objects

**Comment by Phillipus** [comment]
So we need to narrow it down to what version of Archi this problem is in, and perhaps whether it's a configuration issue for your setup.

- Try Archi 4.8.1
- Try a different machine/configuration
- Any other Archi users have this problem?
- Is it an Eclipse version issue?
- Is it something we broke? (for example, we use our own GEF / Draw2d libs)

**Comment by Phillipus** [comment]
Help me to understand this better...

As you type in the diagram text box the direction is correct? It's only when you finish editing that the displayed direction is wrong?

But please try to narrow it down to when this bug started.

**Comment by Phillipus** [comment]
And does this happen in Note text box?

**Comment by Alon-fnx** [comment]
When narrowing the box for one word at a line, It is OK. It shows the words as they ere typed.
In note shape It is the same - when expanded it's reversed, and when narrowing the box it is shown OK..

**Comment by Phillipus** [comment]
What do you mean "narrowing the box"? Do you mean when the edit box is shown?

Please try all controls - Notes, Group, Stickies, Canvas blocks, stickies etc.

**Comment by Alon-fnx** [comment]
"Narrowing the box - i.e. changing the box width

**Comment by Phillipus** [comment]
> "Narrowing the box - i.e. changing the box width

Now I'm even more confused. Are you saying that the direction of text changes when you resize a diagram box? Is it correct? Not correct? When you edit a box? Or what? Please provide the exact steps to reproduce this.

Also, please try  Archi 4.8.1. Without knowing when this issue started I'm having a hard time to help.

**Comment by Alon-fnx** [comment]
![22](https://user-images.githubusercontent.com/97157181/148211521-a53b9c5d-62a6-404e-93d2-a4402982185f.jpg)
![44](https://user-images.githubusercontent.com/97157181/148211526-18600a39-0fc4-4bd3-ae5c-4d251c71be81.jpg)
![33](https://user-images.githubusercontent.com/97157181/148211527-07a2fb72-41a5-4a82-9e9b-0a813c643fed.jpg)
![11](https://user-images.githubusercontent.com/97157181/148211529-9fd1b8a0-71aa-4edc-a175-5ede341bf388.jpg)

**Comment by Phillipus** [comment]
And please answer this question:

> As you type in the diagram text box the direction is correct? It's only when you finish editing that the displayed direction is wrong?

**Comment by Alon-fnx** [comment]
Correct.
Edit pane is OK,

**Comment by Phillipus** [comment]
And please can you answer the questions for Archi versions:

Archi 4.7.1 - works OK? Not OK?
Archi 4.8.1 - works OK? Not OK?
Archi 4.9.2 beta - works OK? Not OK?

**Comment by Alon-fnx** [comment]
If I recall, the last version I had installed was 4.7.1 .
In this version it worked OK.

**Comment by Phillipus** [comment]
Please test 4.8.1 and 4.9.2 beta

**Comment by Alon-fnx** [comment]
I'd love to, but I'm working in an enterprise that every installation can take days... :(
I'll try my best

**Comment by Phillipus** [comment]
> I'd love to, but I'm working in an enterprise that every installation can take days... :(
I'll try my best

In that case I can't provide a fix in a beta version, because you wouldn't be able to test the beta version.

**Comment by Phillipus** [comment]
I've made a test build for Windows on the basis of my theory about a breaking change above:

https://www.archimatetool.com/downloads/beta/Archi-Win64-4.9.2-test1.zip

If it is possible for someone to test this and if it works I can apply the fix.

**Comment by Alon-fnx** [comment]
> I've made a test build for Windows on the basis of my theory about a breaking change above:
> 
> https://www.archimatetool.com/downloads/beta/Archi-Win64-4.9.2-test1.zip
> 
> If it is possible for someone to test this and if it works I can apply the fix.
> 
> **But without user feedback and testing I am not able to fix or diagnose this issue - all I simply need to know is which version of Archi this worked and which version it broke.**

Phillip,
I'm admired by the agility and the swiftness of your response.
I already started to move the wheels.

Thank you - I really do.

**Comment by Alon-fnx** [comment]
My other machine is a Linux...
I hope tomorrow I'll have it installed.

**Comment by Phillipus** [comment]
> If I recall, the last version I had installed was 4.7.1 . In this version it worked OK.

If that is the case then my proposed fix won't make any difference.

This really needs more eyes on the case.

We've not had any other reports about this. For me, as I said, I need to know when it stopped working. Also, need to know if this is an isolated case.

**Comment by Alon-fnx** [comment]
> > If I recall, the last version I had installed was 4.7.1 . In this version it worked OK.
> 
> If that is the case then my proposed fix won't make any difference.
> 
> This really needs more eyes on the case.
> 
> We've not had any other reports about this. For me, as I said, I need to know when it stopped working. Also, need to know if this is an isolated case.

Tested the zip on my home pc (win10).
Same problem...

**Comment by Phillipus** [comment]
> Tested the zip on my home pc (win10). Same problem...

OK, then I need to know the version where it did work. Can you test some versions please:

https://www.archimatetool.com/archive/

**Comment by Alon-fnx** [comment]
4.8.1 no ok
4.7.1 **OK**

**Comment by Phillipus** [comment]
> 4.8.1 no ok
> 4.7.1 **OK**

Thanks, that helps.

So we can eliminate these changes:

- Not using ScaledGraphics class
- Using our own fork of GEF/Draw2d

I reviewed the changes between Archi 4.7.1 and 4.8.1 and I can't see anything that might affect paragraph layout and Bidi support in the diagram figures.

What did change was Archi 4.8.1 uses Eclipse 4.18 instead of 4.15. If it is an Eclipse issue then it might be hard to pinpoint.

Is the issue present on the Mac and Linux versions?

Also, it's strange that this has not been reported before.

**Comment by Alon-fnx** [comment]
I haven't tried on Mac or Linux.
Maybe I'll revert to 4.7.1 in the meantime.
Yes, It's strange that it wasn't reported. I'm sure there are many RTL users.

**Comment by Phillipus** [comment]
Let's assume that the issue is caused by some bug in Eclipse. In order to figure out what's wrong requires somebody with a RTL setup to debug and try some things in the Archi code. I can't do that because I can't set text to Hebrew or Arabic and set things to use RTL.

If someone is able to [set up a test environment](https://github.com/archimatetool/archi/wiki/Developer-Documentation) they could look at the following classes:

```
com.archimatetool.editor.diagram.figures.AbstractTextControlContainerFigure
com.archimatetool.editor.diagram.figures.diagram.NoteFigure
```

For example, in the `NoteFigure` class the text control is set up as follows:

```java
protected void setUI() {
    setLayoutManager(new GridLayout());

    FlowPage page = new FlowPage();
    fTextFlow = new TextFlow();
    fTextFlow.setLayoutManager(new ParagraphTextLayout(fTextFlow, ParagraphTextLayout.WORD_WRAP_SOFT));
    page.add(fTextFlow);
    setOpaque(true);
    
    GridData gd = new GridData(SWT.LEFT, SWT.TOP, true, true);
    add(page, gd);
    
    fTextPositionDelegate = new TextPositionDelegate(this, page, getDiagramModelObject());
    
    setIconicDelegate(new IconicDelegate(getDiagramModelObject()));
}

The `FlowPage` class handles `Bidi` information. One can manually set the RTL like this:

```
page.setOrientation(SWT.RIGHT_TO_LEFT);
```

- Perhaps the top level parent Canvas is not being set to RTL and is not propagating down to child components?
- Perhaps manually setting orientation to RTL might force RTL text?

Just some ideas...

**Comment by Alon-fnx** [comment]
Hello Phillip,

I manged to change the ini file with -dir rtl **and it works O.K.!**
It change the entire ui to RTL but I can adapt to that.

Thank you very much for your time!

Alon

**Comment by Phillipus** [comment]
I still thinks there's a bug in Eclipse:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=566914

**Comment by Alon-fnx** [comment]
> I still thinks there's a bug in Eclipse:
> 
> https://bugs.eclipse.org/bugs/show_bug.cgi?id=566914

Please let me know if the bug is resolved.

**Comment by Phillipus** [comment]
@Alon-fnx 

Perhaps you can help to test a special build of Archi that is built using Eclipse 4.15?

As you say that RTL is working in Archi 4.7.1 which is built on Eclipse 4.15.

It is a buggy build, errors will be written to the log, and it will leak memory...but it will show whether the RTL bug is in Eclipse or not:

https://www.archimatetool.com/downloads/beta/Archi-Win64-4.9.2-testbuild.zip

Perhaps you could launch it and do a quick test? 

Don't use it for production work!

**Comment by Alon-fnx** [comment]
@Phillipus  I'll try

**Comment by Alon-fnx** [comment]
> @Alon-fnx
> 
> Perhaps you can help to test a special build of Archi that is built using Eclipse 4.15?
> 
> As you say that RTL is working in Archi 4.7.1 which is built on Eclipse 4.15.
> 
> It is a buggy build, errors will be written to the log, and it will leak memory...but it will show whether the RTL bug is in Eclipse or not:
> 
> https://www.archimatetool.com/downloads/beta/Archi-Win64-4.9.2-testbuild.zip
> 
> Perhaps you could launch it and do a quick test?
> 
> Don't use it for production work!

Hi Phil,

The Hebrew is correct.

Does it mean it's an Eclipse bug?

**Comment by Phillipus** [comment]
@Alon-fnx Thanks for testing that build. 

> Does it mean it's an Eclipse bug?

Yes, I think so.

The test build uses all of the code as used in Archi 4.9.2 (apart from having to change the calls to Eclipse's `Color`object for backward compatibility). The only difference is that it is built using Eclipse 4.15 (as used in Archi 4.7.1).

This tells me that there's nothing in Archi's code that can be changed and that we should look to this possible bug:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=566914

**Comment by Phillipus** [comment]
...and for the Eclipse developers to even look at this Eclipse bug they will need a reproducible test case in Eclipse itself.

**Comment by Alon-fnx** [comment]
Thank you.
Please let me know whenever you hear it's done.

**Comment by Phillipus** [comment]
@Alon-fnx A new Eclipse bug has been reported:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=579626

Do you think this could be the cause?

**Comment by Phillipus** [comment]
That Eclipse bug fix that might have caused this is for Windows only. @Alon-fnx Can you check whether RTL works in Archi on Linux and Mac?

**Comment by Phillipus** [comment]
I think it's likely that the Eclipse bug at [579626](https://bugs.eclipse.org/bugs/show_bug.cgi?id=565526) could be the cause. The affected Eclipse class was `TextLayout` and Draw2d uses `TextLayout` in its `BidiProcessor` class for RTL text.

**Comment by Phillipus** [comment]
@Alon-fnx I've made a test build of Archi 4.9.3 with a modified `TextLayout` class. Can you test it and see if it fixes the RTL issue?

https://www.archimatetool.com/download/beta/Archi-Modified-4.9.3.zip

**Comment by Alon-fnx** [comment]
Hello @Phillipus,

Thanks for your concern.
I'll try the modified version and report.
I don't have access to a mac or Linux,

**Comment by Alon-fnx** [comment]
Hi,

Got a 404 error on that link

**Comment by Phillipus** [comment]
> Hi,
> 
> Got a 404 error on that link

The link is https://www.archimatetool.com/downloads/beta/Archi-Modified-4.9.3.zip

**Comment by Phillipus** [comment]
@Alon-fnx You probably are getting these GH messages via email. Please note that messages may be edited since, so it's better to look at the message here in GH in a Browser.

**Comment by Alon-fnx** [comment]
@Phillipus, Works like MAGIC!

**Comment by Phillipus** [comment]
@Alon-fnx Thanks for testing. This confirms the Eclipse regression.

The fix involves removing the code that was added in https://bugs.eclipse.org/bugs/show_bug.cgi?id=565526:

https://git.eclipse.org/c/platform/eclipse.platform.swt.git/commit/?id=bd786bc001096a195a4180b432b5d23d1cfcf3a2

What I did was remove those lines, re-compile the `TextLayout` class for Windows, and re-add it to the `org.eclipse.swt.win32.win32.x86_64_3.119.0.v20220223-1102.jar` file which is in Archi's `plugins` folder.

For anyone else reading this affected by this bug I have uploaded this file here:

https://www.archimatetool.com/downloads/beta/org.eclipse.swt.win32.win32.x86_64_3.119.0.v20220223-1102.jar

All you need do is replace this file in Archi's `plugins` folder. Note, this only affects Windows.

Now we need to get the Eclipse people to fix it. So if anyone wishes to add their weight to this:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=579626

**Comment by Phillipus** [comment]
> That Eclipse bug fix that might have caused this is for Windows only. @Alon-fnx Can you check whether RTL works in Archi on Linux and Mac?

Easy to test by copying and pasting this example text into a Figure or a Documentation or Note field:

جارِ التحميل

**Comment by Alon-fnx** [comment]
I don't have an available Mac.
I can try on linux on Wednesday.
The example text isn't required. Hebrew has the same properties.

**Comment by Phillipus** [comment]
> I don't have an available Mac. I can try on linux on Wednesday.

No need to test, thanks anyway. We now know it's Windows only.

> The example text isn't required. Hebrew has the same properties.

This text is the only way I and others can reproduce it simply.

**Comment by Phillipus** [comment]
This is being addressed: https://github.com/eclipse-platform/eclipse.platform.swt/issues/37

**Comment by Phillipus** [comment]
This has now been fixed in Eclipse. Assuming that there are no other side-effects it will make its way into the next version of Eclipse (4.24) in June 2022.

For Archi, I can build a new version based on Eclipse 4.24 after June providing it doesn't contain any other regressions. In the meantime, I suggest affected users download the jar file in https://github.com/archimatetool/archi/issues/796#issuecomment-1094225858 and replace it.

**Comment by Phillipus** [comment]
This has not been entirely fixed. RTL text is not rendered correctly if non-RTL and RTL text is mixed with non-RTL text on the left:

`Some mixed text جارِ التحميل`

See https://github.com/eclipse-platform/eclipse.platform.swt/issues/37#issuecomment-1096399234

This is being tacked here:

https://github.com/eclipse-platform/eclipse.platform.swt/issues/44

**Comment by Alon-fnx** [comment]
This is not a new issue.
I have this problem from the first day using Archi.

**Comment by Phillipus** [comment]
> This has not been entirely fixed. RTL text is not rendered correctly if non-RTL and RTL text is mixed with non-RTL text on the left:
> 
> `Some mixed text جارِ التحميل`
> 
> See [eclipse-platform/eclipse.platform.swt#37 (comment)](https://github.com/eclipse-platform/eclipse.platform.swt/issues/37#issuecomment-1096399234)
> 
> This is being tacked here:
> 
> [eclipse-platform/eclipse.platform.swt#44](https://github.com/eclipse-platform/eclipse.platform.swt/issues/44)

They fixed that now. 😄

**Comment by Alon-fnx** [comment]
Hello @Phillipus,

Are all RTL problems solved in Archi 4.9.3?

Regards,

Alon

**Comment by Phillipus** [comment]
I thought I made this clear in https://github.com/archimatetool/archi/issues/796#issuecomment-1095286376

**Comment by Alon-fnx** [comment]
Thank you

**Comment by Phillipus** [comment]
This is fixed in Eclipse 4.25 and Archi 4.10 is built on Eclipse 4.25. So close this.
